logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -193,7 +208,8 @@ private void innerWriteEntry(DLQEntry entry) throws IOException {
         int eventPayloadSize = RECORD_HEADER_SIZE + record.length;
         if (currentQueueSize.longValue() + eventPayloadSize > maxQueueSize) {
             if (storageType == QueueStorageType.DROP_NEWER) {
-                logger.error("cannot write event to DLQ(path: " + this.queuePath + "): reached maxQueueSize of " + maxQueueSize);
+                lastError = String.format("Cannot write event to DLQ(path: %s): reached maxQueueSize of %d", queuePath, maxQueueSize);
+                droppedEvents.add(1L);
```

Comment:
I think there needs to be a distinction between `dropped` and `discarded`, and I wonder whether we need to count the `discarded` events at all, particularly given the complexity of doing so.

The `dropped` case is always an error that we should be logging, and that the user should be aware of - the an event that will never be added to the DLQ due to the DLQ being full, and the storage policy dropping newer events, this is something we need to warn about and we should be counting.

Events that are `discarded` are removed from the end of the DLQ when it is full, but due to the nature of the DLQ, it is not clear that the events have been processed or not, without co-ordination with the reader - all events may be processed by the DLQ, but eventually, old segments will be deleted if the queue continues to receive events.

Reply:
> I think there needs to be a distinction between `dropped` and `discarded`

Good point, `dropped` are events that are **never** be processed while in `discarded` the events **may** be processed. 

> The `dropped` case is always an error that we should be logging,

We can log and use a specific error message metric ( e.g. 'dropped_last_error` to distinguish from `discarded_last_error`), WDYT?

> Events that are discarded are removed from the end of the DLQ when it is full...

Actually the read events aren't removed, but one development, is to remove consumed segments from reader. In that case the `discarded` metric would assume the semantic of "messages removed and not consumed" so at that point it would match the `dropped` semantic. We could hold on this part, and implement it when the consumer removes segments, WDYT?

Reply:
> We can log and use a specific error message metric ( e.g. 'dropped_last_error` to distinguish from `discarded_last_error`), WDYT?
I'm not sure we need to mark the "error" when discarding events, but it should be reserved for errors that cause us to be unable to write to the DLQ, such as this case here.

> Actually the read events aren't removed, but one development, is to remove consumed segments from reader. In that case the discarded metric would assume the semantic of "messages removed and not consumed" so at that point it would match the dropped semantic. We could hold on this part, and implement it when the consumer removes segments, WDYT?

Yes, I think once we have the code in place that allows us to make the distinction that events have been dropped and not processed, we should mark those as dropped, and in that case, I don't think we need to track `discarded` given the complexity of doing so. WDYT?

Reply:
In that case, distinguish between `discarded` and `dropped` is difficult, because would require reader and writer synchronization between 2 processes. The `discarded` is useless as metric, because doesn't say which fate the messages had.
So as step forward I can remove the count of `discarded`messages, keeping the counter of `dropped` only for messages not inserted into DLQ when `drop_newer` is selected.
How it sounds to you?